### PR TITLE
Properly set EMAR for Interprocedural Optimization

### PR DESCRIPTION
**Pull request by Skylion007**
This properly sets the EMAR for InterproceduralOptimization as described here: https://github.com/emscripten-core/emscripten/issues/11143 and here https://github.com/mosra/magnum/issues/490

### CODE REVIEW

**Comment by Squareys** [comment]
Looks great, thanks @Skylion007 !

@mosra FYI, I do the -flto for Wonderland Engine, too (though, I set those at project level, not for compiling magnum). And faintly remember setting em-ar somewhere also.

File: generic/Emscripten-wasm.cmake [hunk 1]
```diff
@@ -57,6 +57,10 @@ set(CMAKE_CXX_COMPILER "${EMSCRIPTEN_PREFIX}/em++${EMCC_SUFFIX}")
 # and doing it everywhere.
 set(CMAKE_AR "${EMSCRIPTEN_PREFIX}/emar${EMCC_SUFFIX}" CACHE PATH "Path to Emscripten ar")
 set(CMAKE_RANLIB "${EMSCRIPTEN_PREFIX}/emranlib${EMCC_SUFFIX}" CACHE PATH "Path to Emscripten ranlib")
+set(CMAKE_C_COMPILER_AR "${CMAKE_AR}")
+set(CMAKE_CXX_COMPILER_AR "${CMAKE_AR}")
+set(CMAKE_C_COMPILER_RANLIB "${CMAKE_RANLIB}")
+set(CMAKE_CXX_COMPILER_RANLIB "${CMAKE_RANLIB}")
```
Comment (Squareys) [inline generic/Emscripten-wasm.cmake:63]: I would have suspected that CMake initializes these automatically based on CMAKE_AR and CMAKE_RANLIB right after the toolchain file, but I assume that is not the case?

File: generic/Emscripten-wasm.cmake [hunk 2]
```diff
@@ -76,5 +80,5 @@ set(CMAKE_SYSTEM_PREFIX_PATH ${CMAKE_FIND_ROOT_PATH})
 cmake_minimum_required(VERSION 3.7)
 set(CMAKE_CXX_FLAGS_INIT "-s WASM=1")
 set(CMAKE_EXE_LINKER_FLAGS_INIT "-s WASM=1")
-set(CMAKE_CXX_FLAGS_RELEASE_INIT "-DNDEBUG -O3")
-set(CMAKE_EXE_LINKER_FLAGS_RELEASE_INIT "-O3 --llvm-lto 1")
+set(CMAKE_CXX_FLAGS_RELEASE_INIT "-DNDEBUG -O3 -flto")
+set(CMAKE_EXE_LINKER_FLAGS_RELEASE_INIT "-O3 --llvm-lto 1 -flto")
```
Comment (Squareys) [inline generic/Emscripten-wasm.cmake:84]: @mosra For your reference, here's the documentation for that: https://emscripten.org/docs/compiling/WebAssembly.html#backends

`--llvm-lto 1` is for the deprecated fastcomp backend (removed in emsc 2.0.0), `-flto(=full|thin)` is the new way of achieving link-time optimization

Comment (Squareys) [reply generic/Emscripten-wasm.cmake:84]: Ah, found https://github.com/mosra/toolchains/issues/13, so you are already aware

Comment (Skylion007) [reply generic/Emscripten-wasm.cmake:84]: @Squareys Yeah, unfortunately @mosra mentioned we need to support the fastcomp backend. Do you know an easy to way to check if we are using fastcomp or the more modern clang?

File: generic/Emscripten-wasm.cmake (hunk 1 above)
Comment (Skylion007) [reply generic/Emscripten-wasm.cmake:63]: @Squareys I wish. Emscriptem's toolchain does properly handle it, but Magnum's doesn't sadly.

File: generic/Emscripten-wasm.cmake (hunk 2 above)
Comment (Skylion007) [reply generic/Emscripten-wasm.cmake:84]: Also I was having trouble with the default build running into issues with ThinLTO. @Squareys have you gotten ThinLTO working with build? I got complaints from the wasm-ld.

**Comment by Skylion007** [comment]
Ping @mosra

**Comment by mosra** [comment]
This was _finally_ merged in 98bdb68cb17d9c70d4ba179509075ba479764a5b and b25ca8905d48a5d380eda95dd1a299216a3cad78. The extra delay allowed me to not have to care about fastcomp compatibility anymore.

**Pull request closed without merging**